File: api/internal/contact/validate.go
package contact

import (
	"strings"
	"time"
	"unicode/utf8"

	"github.com/G1NG4R/timseil-dev/api/internal/httpx"
	"github.com/G1NG4R/timseil-dev/api/internal/mail"
)

// verdict is what validation decides. Three outcomes, not two, because the
// build plan counts five answer paths and one of them is a refusal that must
// not look like one.
type verdict int

const (
	// accepted: store it and try to send it.
	accepted verdict = iota

	// rejected: 400 with one invalidParams entry per field.
	rejected

	// discarded: 202 with a receipt that leads nowhere. No row, no mail.
	//
	// The honeypot and the dwell floor both end here, and the contract uses the
	// same word for both: `company` "must be empty; anything else is discarded
	// silently", `dwellMs` "below 3000 is discarded". A 400 would tell a bot
	// which rule caught it, and it would be right once.
	discarded
)

// submission is a validated body: every field trimmed, every bound checked, and
// the two hashes' inputs settled. Nothing downstream re-checks it, so nothing
// downstream needs to know how a visitor spells things.
type submission struct {
	name     string
	email    string
	message  string
	dwellMs  int32
	clientTs time.Time
}

// validate applies every rule the contract states and two it does not.
//
// The first extra is the CRLF refusal. The contract bounds the length of an
// address and calls it `format: email`, neither of which stops a `\r\n` — and
// the address becomes a Reply-To header, where a line break is a second header
// written by a stranger. The name is checked for the same reason: it becomes
// the Subject. Appendix F of the build plan lists both.
//
// The second is that the address must survive net/mail's parser as a bare
// addr-spec. `format: email` is annotation, not validation; a generated client
// will happily send `"x" <y@z>` and every OpenAPI validator will pass it.
//
// The order of the returned params follows the order of the form, so the page
// can move focus to the first failing field without sorting anything.
func validate(body httpx.ContactRequest, now time.Time) (submission, verdict, []httpx.InvalidParam) {
	// The two silent refusals come first and short-circuit everything. There is
	// no point telling a bot that its message was also too short, and there is
	// no point spending a database round trip on it.
	if discardReason(body) != "" {
		return submission{}, discarded, nil
	}

	var invalid []httpx.InvalidParam
	fail := func(name, reason string) {
		invalid = append(invalid, httpx.InvalidParam{Name: name, Reason: reason})
	}

	name := strings.TrimSpace(body.Name)
	switch {
	case containsControl(name):
		// Never echoed back. The reason a visitor sees is about the shape of
		// what they sent, never a copy of it — a problem document that quotes
		// attacker input is a reflection vector looking for a renderer.
		fail("name", "must not contain line breaks or control characters")
	case utf8.RuneCountInString(name) < minName:
		fail("name", "at least 2 characters")
	case utf8.RuneCountInString(name) > maxName:
		fail("name", "at most 80 characters")
	}

	email := strings.TrimSpace(string(body.Email))
	switch {
	case email == "":
		fail("email", "required")
	case len(email) > maxEmail:
		// Counted in bytes, not runes: 254 is the octet limit of an address in
		// RFC 5321, and the database column agrees with it in octets.
		fail("email", "at most 254 characters")
	case !deliverable(email):
		// net/mail accepts `anna@example`, and it is right to: a domain without
		// a dot is a valid addr-spec and `root@localhost` is a real address on
		// somebody's machine. It is not a real address on the internet, and the
		// only thing this form does with one is put it in a Reply-To that
		// bounces. The contact page applies the same rule before it sends.
		//
		// One reason for every way an address can be wrong, and the reason never
		// quotes what was sent.
		fail("email", "not a plain mail address")
	}

	message := strings.TrimSpace(body.Message)
	switch {
	case utf8.RuneCountInString(message) < minMessage:
		fail("message", "at least 20 characters")
	case utf8.RuneCountInString(message) > maxMessage:
		fail("message", "at most 4000 characters")
	}

	// A timestamp far outside any plausible clock is not a double-click key, it
	// is something to keep out of a timestamptz column. See maxClockSkew for why
	// the window is loose rather than tight.
	if body.Ts.IsZero() || absDuration(now.Sub(body.Ts)) > maxClockSkew {
		fail("ts", "not a plausible timestamp")
	}

	// The column is `integer`. A body that overflows it would be refused by
	// Postgres with a message nobody can act on, three layers away from here.
	if body.DwellMs > maxInt32 {
		fail("dwellMs", "implausibly large")
	}

	if len(invalid) > 0 {
		return submission{}, rejected, invalid
	}

	return submission{
		name:     name,
		email:    email,
		message:  message,
		dwellMs:  int32(body.DwellMs),
		clientTs: body.Ts.UTC(),
	}, accepted, nil
}

// discardReason names the silent rule that caught a submission, or "" if none
// did. It is the rule itself and not a copy of it: the log line that says which
// trap fired asks this function, so the two cannot come to disagree about what
// a honeypot is.
//
// company is compared untrimmed. A field a human never sees should be exactly
// the empty string, and a browser that sends a space to a hidden input is doing
// something worth refusing.
func discardReason(body httpx.ContactRequest) string {
	switch {
	case body.Company != "":
		return "honeypot"
	case body.DwellMs < minDwell:
		return "dwell"
	default:
		return ""
	}
}

const maxInt32 = 1<<31 - 1

// deliverable holds an address to two rules: net/mail's, and the one net/mail
// has no opinion about.
//
// The first is mail.BareAddress — an addr-spec with no display name, no second
// recipient and no line break. It is the same function internal/config applies
// to SMTP_USERNAME and internal/mail applies again at send time, so there is one
// answer to "what is an address" rather than three.
//
// The second is that the domain has a dot and a last label that looks like a
// public suffix. That is not RFC 5322, deliberately: what this endpoint needs is
// an address a reply can reach, and the two are not the same question.
func deliverable(address string) bool {
	if _, err := mail.BareAddress(address); err != nil {
		return false
	}

	at := strings.LastIndex(address, "@")
	domain := address[at+1:]

	// A bracketed address literal — anna@[192.0.2.1] — parses and is
	// technically deliverable, and is also never what a visitor meant to type.
	dot := strings.LastIndex(domain, ".")
	if dot <= 0 || dot == len(domain)-1 {
		return false
	}

	tld := domain[dot+1:]
	if len(tld) < 2 {
		return false
	}
	for _, r := range tld {
		if !(r >= 'a' && r <= 'z') && !(r >= 'A' && r <= 'Z') {
			// Punycode, so an internationalised domain arrives here as xn--…
			// and passes; anything else is not a top-level domain.
			return false
		}
	}
	return true
}

// containsControl refuses anything that is not printable text.
//
// Wider than a CR/LF check, and cheaply so. What has to be kept out of a header
// is a line break, but U+2028, U+2029 and NUL are line breaks or terminators to
// some parser somewhere, and a name containing a bare escape character is not a
// name. mail.Build refuses these a second time; this is the layer that can say
// which field it was.
func containsControl(s string) bool {
	for _, r := range s {
		if r == '\u2028' || r == '\u2029' {
			return true
		}
		// C0 and C1, tab included: a tab in a Subject folds the header.
		if r < 0x20 || (r >= 0x7f && r <= 0x9f) {
			return true
		}
	}
	return false
}

func absDuration(d time.Duration) time.Duration {
	if d < 0 {
		return -d
	}
	return d
}

File: web/lib/contact/validate.ts
// The client's half of the contact validator, and it is a MIRROR rather than an
// authority.
//
// `api/internal/contact/validate.go` is the authority and stays it. This file
// exists for one reason the api cannot serve: a visitor who mistypes an address
// should learn it beside the field, not after a round trip that also spends one
// of their three sends in ten minutes. Everything here is therefore checked
// again on the other side, and the page renders the api's answer when the two
// disagree.
//
// SO IT LEANS LOOSE, ON PURPOSE, AND ONLY IN ONE DIRECTION. A client rule
// stricter than the server's refuses an address the api would have accepted, and
// the visitor has no way to argue with it. A client rule looser than the
// server's costs a round trip and shows the server's reason under the same
// field. `deliverable` below is the place this matters: Go runs net/mail's
// parser and this cannot, so it checks the shape that catches typos and lets the
// remainder travel.
//
// THE REASONS ARE COPIED WORD FOR WORD from validate.go, and the order of the
// returned entries follows FIELDS, which is the order of the form. That is the
// promise validate.go:53-55 makes to this page — "so the page can move focus to
// the first failing field without sorting anything" — and it is only a promise
// kept if both sides read the same way down.

import type { components } from "../api/schema";

import { FIELDS, lengthOf } from "./fields.ts";

/** One entry of the contract's `invalidParams`, so the page renders an api
 *  answer and a local one through the same branch. */
export type InvalidParam = components["schemas"]["InvalidParam"];

/** What the visitor has typed, before anything is built out of it. */
export interface Draft {
  readonly name: string;
  readonly email: string;
  readonly message: string;
}

/**
 * Refuses anything that is not printable text.
 *
 * `containsControl` in validate.go, rune for rune: U+2028 and U+2029, plus C0
 * and C1 including tab. A tab in a Subject folds the header, which is the whole
 * reason the name is checked at all.
 */
export function containsControl(value: string): boolean {
  for (const rune of value) {
    if (rune === "\u2028" || rune === "\u2029") return true;
    const code = rune.codePointAt(0) ?? 0;
    if (code < 0x20 || (code >= 0x7f && code <= 0x9f)) return true;
  }
  return false;
}

/**
 * Whether a reply could reach this address.
 *
 * The second half of Go's `deliverable`, which is the half a typo trips over: a
 * domain with a dot, a last label of at least two letters, and nothing that is
 * obviously not an address. The first half is net/mail's parser, and the api
 * keeps it.
 *
 * PUNYCODE PASSES because an internationalised domain arrives as `xn--…`, which
 * is ASCII letters and digits — the same reason Go's loop gives.
 */
export function deliverable(address: string): boolean {
  if (containsControl(address)) return false;
  if (/[\s<>,;"]/.test(address)) return false;

  const at = address.lastIndexOf("@");
  if (at <= 0 || at === address.length - 1) return false;
  if (address.slice(0, at).includes("@")) return false;

  const domain = address.slice(at + 1);
  const dot = domain.lastIndexOf(".");
  if (dot <= 0 || dot === domain.length - 1) return false;

  const tld = domain.slice(dot + 1);
  if (tld.length < 2) return false;
  return /^[A-Za-z]+$/.test(tld);
}

/**
 * Every rule the form can check, in the order of the form.
 *
 * Returns `[]` when there is nothing to say. It does not check `dwellMs`, `ts`
 * or `company`: those three are not typed by anybody, lib/contact/payload.ts
 * builds them, and a form that told a visitor their honeypot was wrong would be
 * describing a bug rather than a mistake.
 */
export function validateDraft(draft: Draft): readonly InvalidParam[] {
  const invalid: InvalidParam[] = [];

  for (const field of FIELDS) {
    // Trimmed first, exactly as validate.go does: a message of twenty spaces is
    // not twenty characters, and a visitor should learn that here rather than
    // from a 400.
    const value = draft[field.name].trim();
    const used = lengthOf(value, field.unit);

    if (field.name === "name") {
      if (containsControl(value)) {
        invalid.push({ name: "name", reason: "must not contain line breaks or control characters" });
        continue;
      }
    }

    if (field.name === "email") {
      if (value === "") {
        invalid.push({ name: "email", reason: "required" });
        continue;
      }
      if (used > field.max) {
        invalid.push({ name: "email", reason: "at most 254 characters" });
        continue;
      }
      if (!deliverable(value)) {
        invalid.push({ name: "email", reason: "not a plain mail address" });
      }
      continue;
    }

    if (field.min !== null && used < field.min) {
      invalid.push({ name: field.name, reason: `at least ${String(field.min)} characters` });
      continue;
    }
    if (used > field.max) {
      invalid.push({ name: field.name, reason: `at most ${String(field.max)} characters` });
    }
  }

  return invalid;
}

/**
 * The first field with something wrong with it, or `null`.
 *
 * A separate function because the api answers with the same list and the page
 * moves focus the same way for both. `invalidParams` may name a field this form
 * does not draw — `ts` and `dwellMs` are in the contract — and a focus call
 * against a field that is not on the screen would silently do nothing, so the
 * lookup goes through FIELDS rather than through the list.
 */
export function firstInvalidField(invalid: readonly InvalidParam[]): string | null {
  for (const field of FIELDS) {
    if (invalid.some((entry) => entry.name === field.name)) return field.name;
  }
  return null;
}

/** The reason for one field, or `undefined` — which is what `Field` wants. */
export function reasonFor(invalid: readonly InvalidParam[], name: string): string | undefined {
  return invalid.find((entry) => entry.name === name)?.reason;
}
